### PR TITLE
Update NATS imf file

### PR DESCRIPTION
**Pull request by variadico**
We reorganized the directory layout of [nats-io/nats-docker](https://github.com/nats-io/nats-docker). This change
reflects the changes from that repo.

cc @kozlovic @wallyqs

### CODE REVIEW

**Comment by variadico** [comment]
Commit d56a6de failed because:

```
$ ./bashbrew/travis.sh
...
failed fetching repo "golang:1.12.9"
tag not found in manifest for "golang": "1.12.9"
```

In [nats-io/nats-docker](https://github.com/nats-io/nats-docker/blob/master/2.1.0/scratch/Dockerfile#L2), we use `golang:1.12.9` as a build image for our final NATS image. Does this error mean that we must use a Go tag that currently exists in the `golang` imf file? 

Meaning, do we have to update our Dockerfile to use `golang:1.12.10` instead of `golang:1.12.9`?

**Comment by yosifkit** [comment]
> Meaning, do we have to update our Dockerfile to use golang:1.12.10 instead of golang:1.12.9?

Yes, we only update the golang images that contains the versions that upstream Go supports. So "old" versions like `1.12.9` are no longer updated.

> 2. two-stage build where the necessary **artifact does not exist and must be built from source** ...
>
>    - a Go project without official binary releases (although it is highly recommended for something trivial like Go to publish actual official release binaries, especially if the Go version required/supported for building is highly specific given that Go only supports two major releases at a time)
>
> \- https://github.com/docker-library/faq/tree/b2c55617dc6d924892b4d2599765763915a246aa#multi-stage-builds (emphasis added)

Unfortunately the [new reorganization](https://github.com/nats-io/nats-docker/commit/704a6264de48ec1a3720c475ac23d35ccf03921a#diff-0b567411b70abca14ccd2cc2e2c5b7ee) doesn't fit within our allowed mulit-stage builds (quoted text above); I'd recommend a few alternatives:

1. if the specific go version is required/important, then using the release binaries from https://github.com/nats-io/nats-server/releases would be the prefered setup (this would also require making a non `scratch` image that could be copied from if a scratch image is still needed)
1. make the first stage a real image and use `COPY --from=nats:2.1.0-linux` in the `FROM scratch` image

------

> In 816f3d4, I was trying to test this hypothesis, but the test failed because I didn't commit to master. It's kinda unfortunate that I would have to commit to master in order to test a hunch.

That would need a `GitFetch: refs/heads/update-go` (I think) in order to know what branch to fetch that commit from.

**Comment by variadico** [comment]
Awesome, thanks for the help, @yosifkit!

Yeah, my original approach was to simply download the binaries. However, turns out we build the binaries slightly differently for Docker. For example, for the non-Docker release binaries, we do not disable CGO. But, for the scratch image we have to.

However, even if we could use the release binaries. We still need some sort of second step to create the scratch image.

Would it be okay if we updated our build stage to this?

```diff
- FROM golang:1.12.9 as tmp
+ FROM golang:1.12.10 as tmp
```

or maybe

```diff
- FROM golang:1.12.9 as tmp
+ FROM golang:1.12 as tmp
``` 

I'm guessing we would probably have to update the NATS imf whenever Go removes 1.12.10 Dockerfiles. But is that how things work anyway? What happens when Debian removes `buster` Dockerfiles?

Not sure how things work. Thanks again! :D

**Comment by tianon** [comment]
The officially-released `nats-server` binary for `amd64` is ~10MB.  The `alpine:3.10` image is ~5.5MB.

What I'd suggest is simply not being `FROM scratch` anymore:

```dockerfile
FROM alpine:3.10

ENV SERVER_VERSION v2.1.0

RUN set -eux; \
	apkArch="$(apk --print-arch)"; \
	case "$apkArch" in \
		arm64) natsArch='arm64' ;; \
		armhf) natsArch='arm6' ;; \
		armv7) natsArch='arm7' ;; \
		x86) natsArch='386' ;; \
		x86_64) natsArch='amd64' ;; \
		*) echo >&2 "error: $apkArch is not supported!"; exit 1 ;; \
	esac; \
	wget -O nats-server.zip "https://github.com/nats-io/nats-server/releases/download/$SERVER_VERSION/nats-server-$SERVER_VERSION-linux-$natsArch.zip"; \
	apk add --no-cache --virtual .temp unzip; \
	unzip nats-server.zip "nats-server-$SERVER_VERSION-linux-$natsArch/nats-server"; \
	rm nats-server.zip; \
	mv "nats-server-$SERVER_VERSION-linux-$natsArch/nats-server" /; \
	rmdir "nats-server-$SERVER_VERSION-linux-$natsArch"; \
	apk del --no-network .temp

#COPY nats-server.conf /nats-server.conf

CMD ["/nats-server"]
CMD ["--config", "nats-server.conf"]
```

The resulting image is, as expected, ~15.5MB and contains useful utilities like a shell and a package manager, so IMO the deployment flexibility is much greater for your users than a `FROM scratch` image can be.

**Comment by variadico** [comment]
Yep, that's actually what my next PR would be. In addition to scratch, we'd also like to provide customers with an alpine image.

For some customers utilities like a shell and a package manager are liabilities. For others, not. Either way, we'd like to allow our customers to choose whichever flavor best suits their environment.

**Comment by kozlovic** [comment]
@tianon @yosifkit Please do not merge until we figure out internally if we are going with the new approach in nats-docker. Will post here when we have made a decision. Thanks!

**Comment by variadico** [comment]
All right. We're ready for another review, @tianon, @yosifkit. :pray:

On our side, we've implemented your suggestions. https://github.com/nats-io/nats-docker/pull/21

* We got rid of multi-stage builds.
* On Windows, we use servercore to download the binary and copy it to the nanoservers, similar to [how Go does it].

In the near future, we'll be adding Alpine and it'll download the binary, as suggested above.

The scratch container is the weird one of the bunch. For that, we just reverted to the old way of doing things. We download the binaries and check them into our repo. Exactly as before.

[how Go does it]: https://github.com/docker-library/golang/blob/master/1.13/windows/nanoserver-1809/Dockerfile#L19

**Comment by kozlovic** [comment]
> We download the binaries and check them into our repo. Exactly as before.

I think you meant we build the binaries and check them into our repo.

**Comment by kozlovic** [comment]
@yosifkit @tianon Ok, so we are ready for the review. We may release a NATS Server 2.1.1 today or tomorrow, so would be based on the new layout. Thanks!

**Comment by yosifkit** [comment]
<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index 0cd1b90..3172d66 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -2,6 +2,8 @@ nats:linux @ amd64
 nats:linux @ arm32v6
 nats:linux @ arm32v7
 nats:linux @ arm64v8
-nats:nanoserver-1803 @ windows-amd64
-nats:nanoserver-1809 @ windows-amd64
-nats:windowsservercore @ windows-amd64
+nats:nanoserver1803 @ windows-amd64
+nats:nanoserver1809 @ windows-amd64
+nats:servercore1803 @ windows-amd64
+nats:servercore1809 @ windows-amd64
+nats:servercoreltsc2016 @ windows-amd64
diff --git a/_bashbrew-list b/_bashbrew-list
index e532ae7..17b8f1a 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -1,11 +1,14 @@
 nats:2.1.0
 nats:2.1.0-linux
-nats:2.1.0-nanoserver
-nats:2.1.0-nanoserver-1809
-nats:2.1.0-windowsservercore
+nats:2.1.0-nanoserver1803
+nats:2.1.0-nanoserver1809
+nats:2.1.0-servercore1803
+nats:2.1.0-servercore1809
+nats:2.1.0-servercoreltsc2016
 nats:latest
 nats:linux
-nats:nanoserver
-nats:nanoserver-1803
-nats:nanoserver-1809
-nats:windowsservercore
+nats:nanoserver1803
+nats:nanoserver1809
+nats:servercore1803
+nats:servercore1809
+nats:servercoreltsc2016
diff --git a/_bashbrew-shared-tags b/_bashbrew-shared-tags
index 2cc84bd..8248384 100644
--- a/_bashbrew-shared-tags
+++ b/_bashbrew-shared-tags
@@ -1,2 +1,2 @@
-nats:2.1.0 -- nats:linux, nats:nanoserver-1803
-nats:latest -- nats:linux, nats:nanoserver-1803
+nats:2.1.0 -- nats:linux, nats:nanoserver1809
+nats:latest -- nats:linux, nats:nanoserver1809
diff --git a/nats_linux/Dockerfile b/nats_linux/Dockerfile
index a616708..a5d0824 100644
--- a/nats_linux/Dockerfile
+++ b/nats_linux/Dockerfile
@@ -1,11 +1,6 @@
 FROM scratch
-
 COPY nats-server /nats-server
-COPY nats-server.conf nats-server.conf
-
-# Expose client, management, and cluster ports
+COPY nats-server.conf /nats-server.conf
 EXPOSE 4222 8222 6222
-
-# Run via the configuration file
 ENTRYPOINT ["/nats-server"]
-CMD ["-c", "nats-server.conf"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1803/Dockerfile b/nats_nanoserver-1803/Dockerfile
deleted file mode 100644
index 64479ed..0000000
diff --git a/nats_nanoserver-1803/nats-server.exe b/nats_nanoserver-1803/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
diff --git a/nats_nanoserver-1809/Dockerfile b/nats_nanoserver-1809/Dockerfile
deleted file mode 100644
index 465a528..0000000
diff --git a/nats_nanoserver-1809/nats-server.exe b/nats_nanoserver-1809/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
diff --git a/nats_nanoserver1803/Dockerfile b/nats_nanoserver1803/Dockerfile
new file mode 100644
index 0000000..8c9d154
--- /dev/null
+++ b/nats_nanoserver1803/Dockerfile
@@ -0,0 +1,10 @@
+FROM mcr.microsoft.com/windows/nanoserver:1803
+
+ENV NATS_DOCKERIZED 1
+
+COPY --from=nats:2.1.0-servercore1803 C:\\nats-server.exe C:\\nats-server.exe
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1809/nats-server.conf b/nats_nanoserver1803/nats-server.conf
similarity index 99%
rename from nats_nanoserver-1809/nats-server.conf
rename to nats_nanoserver1803/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_nanoserver1809/Dockerfile b/nats_nanoserver1809/Dockerfile
new file mode 100644
index 0000000..b745574
--- /dev/null
+++ b/nats_nanoserver1809/Dockerfile
@@ -0,0 +1,10 @@
+FROM mcr.microsoft.com/windows/nanoserver:1809
+
+ENV NATS_DOCKERIZED 1
+
+COPY --from=nats:2.1.0-servercore1809 C:\\nats-server.exe C:\\nats-server.exe
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1803/nats-server.conf b/nats_nanoserver1809/nats-server.conf
similarity index 99%
rename from nats_nanoserver-1803/nats-server.conf
rename to nats_nanoserver1809/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_servercore1803/Dockerfile b/nats_servercore1803/Dockerfile
new file mode 100644
index 0000000..49987be
--- /dev/null
+++ b/nats_servercore1803/Dockerfile
@@ -0,0 +1,30 @@
+FROM mcr.microsoft.com/windows/servercore:1803
+
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
+
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
+
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
+
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_windowsservercore/nats-server.conf b/nats_servercore1803/nats-server.conf
similarity index 99%
rename from nats_windowsservercore/nats-server.conf
rename to nats_servercore1803/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_servercore1809/Dockerfile b/nats_servercore1809/Dockerfile
new file mode 100644
index 0000000..8bb790f
--- /dev/null
+++ b/nats_servercore1809/Dockerfile
@@ -0,0 +1,30 @@
+FROM mcr.microsoft.com/windows/servercore:1809
+
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
+
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
+
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
+
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_linux/nats-server.conf b/nats_servercore1809/nats-server.conf
similarity index 99%
copy from nats_linux/nats-server.conf
copy to nats_servercore1809/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_servercoreltsc2016/Dockerfile b/nats_servercoreltsc2016/Dockerfile
new file mode 100644
index 0000000..fe9c941
--- /dev/null
+++ b/nats_servercoreltsc2016/Dockerfile
@@ -0,0 +1,30 @@
+FROM mcr.microsoft.com/windows/servercore:ltsc2016
+
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
+
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
+
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
+
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_linux/nats-server.conf b/nats_servercoreltsc2016/nats-server.conf
similarity index 99%
copy from nats_linux/nats-server.conf
copy to nats_servercoreltsc2016/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_windowsservercore/Dockerfile b/nats_windowsservercore/Dockerfile
deleted file mode 100644
index 64c6aaf..0000000
diff --git a/nats_windowsservercore/nats-server.exe b/nats_windowsservercore/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
```

</details>

File: library/nats [hunk 1]
```diff
@@ -2,35 +2,38 @@ Maintainers: Derek Collison <derek@synadia.com> (@derekcollison),
...
-arm32v7-Directory: arm32v7
-arm64v8-GitCommit: 710f0ed18645d78e97fa7fd8cdf9b80dbe936eb6
-arm64v8-Directory: arm64v8
+amd64-Directory: 2.1.0/scratch/amd64
+arm32v6-Directory: 2.1.0/scratch/arm32v6
+arm32v7-Directory: 2.1.0/scratch/arm32v7
+arm64v8-Directory: 2.1.0/scratch/arm64v8
 
-Tags: 2.1.0-nanoserver, nanoserver, nanoserver-1803
+Tags: 2.1.0-servercore1809, servercore1809
```
Comment (yosifkit) [inline library/nats]: To keep consistency with the rest of the official images, these should be `windowsservercore`

Comment (variadico) [reply library/nats]: Updated. :+1:

Comment (variadico) [reply library/nats]: Wait... :(
Oh, I see. We're gonna have to update our repo again.
Okay. One sec.

Comment (variadico) [reply library/nats]: @yosifkit can I get a review of changes on our side before we merge?
https://github.com/nats-io/nats-docker/pull/23

Just want to make sure it matches up with requirements on your end.

Comment (variadico) [reply library/nats]: Aaaaalllll righty. Rename done. CI green. :heavy_check_mark:

Comment (variadico) [reply library/nats]: Friendly ping, @yosifkit. Sorry we keep bothering you. Promise the next PR will go a lot smoother. :smile: :pray:

**Comment by tianon** [comment]
<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index 0cd1b90..16c2181 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -5,3 +5,5 @@ nats:linux @ arm64v8
 nats:nanoserver-1803 @ windows-amd64
 nats:nanoserver-1809 @ windows-amd64
 nats:windowsservercore @ windows-amd64
+nats:windowsservercore-1803 @ windows-amd64
+nats:windowsservercore-ltsc2016 @ windows-amd64
diff --git a/_bashbrew-list b/_bashbrew-list
index e532ae7..73eef51 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -1,11 +1,15 @@
 nats:2.1.0
 nats:2.1.0-linux
-nats:2.1.0-nanoserver
+nats:2.1.0-nanoserver-1803
 nats:2.1.0-nanoserver-1809
-nats:2.1.0-windowsservercore
+nats:2.1.0-windowsservercore-1803
+nats:2.1.0-windowsservercore-1809
+nats:2.1.0-windowsservercore-ltsc2016
 nats:latest
 nats:linux
-nats:nanoserver
 nats:nanoserver-1803
 nats:nanoserver-1809
 nats:windowsservercore
+nats:windowsservercore-1803
+nats:windowsservercore-1809
+nats:windowsservercore-ltsc2016
diff --git a/_bashbrew-shared-tags b/_bashbrew-shared-tags
index 2cc84bd..4e59c13 100644
--- a/_bashbrew-shared-tags
+++ b/_bashbrew-shared-tags
@@ -1,2 +1,2 @@
-nats:2.1.0 -- nats:linux, nats:nanoserver-1803
-nats:latest -- nats:linux, nats:nanoserver-1803
+nats:2.1.0 -- nats:linux, nats:nanoserver-1809
+nats:latest -- nats:linux, nats:nanoserver-1809
diff --git a/nats_linux/Dockerfile b/nats_linux/Dockerfile
index a616708..a5d0824 100644
--- a/nats_linux/Dockerfile
+++ b/nats_linux/Dockerfile
@@ -1,11 +1,6 @@
 FROM scratch
-
 COPY nats-server /nats-server
-COPY nats-server.conf nats-server.conf
-
-# Expose client, management, and cluster ports
+COPY nats-server.conf /nats-server.conf
 EXPOSE 4222 8222 6222
-
-# Run via the configuration file
 ENTRYPOINT ["/nats-server"]
-CMD ["-c", "nats-server.conf"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1803/Dockerfile b/nats_nanoserver-1803/Dockerfile
index 64479ed..4f384c8 100644
--- a/nats_nanoserver-1803/Dockerfile
+++ b/nats_nanoserver-1803/Dockerfile
@@ -1,18 +1,10 @@
 FROM mcr.microsoft.com/windows/nanoserver:1803
 
-# The NAT Server will look for this environment variable.
-# When set, it prevents the use of the service API to detect
-# if it is running in interactive mode or not, which is
-# failing in the context of a Docker container.
-# (https://github.com/nats-io/gnatsd/issues/543)
-ENV NATS_DOCKERIZED=1
+ENV NATS_DOCKERIZED 1
 
-WORKDIR c:/nats-server
-COPY nats-server.exe nats-server.exe
-COPY nats-server.conf nats-server.conf
+COPY --from=nats:2.1.0-windowsservercore-1803 C:\\nats-server.exe C:\\nats-server.exe
+COPY nats-server.conf C:\\nats-server.conf
 
-# Expose client, management, and cluster ports
 EXPOSE 4222 8222 6222
-
-ENTRYPOINT ["nats-server"]
-CMD ["-c", "nats-server.conf"]
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1803/nats-server.exe b/nats_nanoserver-1803/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
diff --git a/nats_nanoserver-1809/Dockerfile b/nats_nanoserver-1809/Dockerfile
index 465a528..68f8d6c 100644
--- a/nats_nanoserver-1809/Dockerfile
+++ b/nats_nanoserver-1809/Dockerfile
@@ -1,18 +1,10 @@
 FROM mcr.microsoft.com/windows/nanoserver:1809
 
-# The NAT Server will look for this environment variable.
-# When set, it prevents the use of the service API to detect
-# if it is running in interactive mode or not, which is
-# failing in the context of a Docker container.
-# (https://github.com/nats-io/gnatsd/issues/543)
-ENV NATS_DOCKERIZED=1
+ENV NATS_DOCKERIZED 1
 
-WORKDIR c:/nats-server
-COPY nats-server.exe nats-server.exe
-COPY nats-server.conf nats-server.conf
+COPY --from=nats:2.1.0-windowsservercore-1809 C:\\nats-server.exe C:\\nats-server.exe
+COPY nats-server.conf C:\\nats-server.conf
 
-# Expose client, management, and cluster ports
 EXPOSE 4222 8222 6222
-
-ENTRYPOINT ["nats-server"]
-CMD ["-c", "nats-server.conf"]
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_nanoserver-1809/nats-server.exe b/nats_nanoserver-1809/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
diff --git a/nats_windowsservercore-1803/Dockerfile b/nats_windowsservercore-1803/Dockerfile
new file mode 100644
index 0000000..49987be
--- /dev/null
+++ b/nats_windowsservercore-1803/Dockerfile
@@ -0,0 +1,30 @@
+FROM mcr.microsoft.com/windows/servercore:1803
+
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
+
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
+
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
+
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_linux/nats-server.conf b/nats_windowsservercore-1803/nats-server.conf
similarity index 99%
copy from nats_linux/nats-server.conf
copy to nats_windowsservercore-1803/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_windowsservercore-ltsc2016/Dockerfile b/nats_windowsservercore-ltsc2016/Dockerfile
new file mode 100644
index 0000000..fe9c941
--- /dev/null
+++ b/nats_windowsservercore-ltsc2016/Dockerfile
@@ -0,0 +1,30 @@
+FROM mcr.microsoft.com/windows/servercore:ltsc2016
+
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
+
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
+
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
+
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_linux/nats-server.conf b/nats_windowsservercore-ltsc2016/nats-server.conf
similarity index 99%
copy from nats_linux/nats-server.conf
copy to nats_windowsservercore-ltsc2016/nats-server.conf
index a94271c..e84ec39 100644
diff --git a/nats_windowsservercore/Dockerfile b/nats_windowsservercore/Dockerfile
index 64c6aaf..8bb790f 100644
--- a/nats_windowsservercore/Dockerfile
+++ b/nats_windowsservercore/Dockerfile
@@ -1,18 +1,30 @@
-FROM mcr.microsoft.com/windows/servercore:ltsc2016
+FROM mcr.microsoft.com/windows/servercore:1809
 
-# The NAT Server will look for this environment variable.
-# When set, it prevents the use of the service API to detect
-# if it is running in interactive mode or not, which is
-# failing in the context of a Docker container.
-# (https://github.com/nats-io/gnatsd/issues/543)
-ENV NATS_DOCKERIZED=1
+# Enable exit on error.
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop';"]
 
-WORKDIR c:/nats-server
-COPY nats-server.exe nats-server.exe
-COPY nats-server.conf nats-server.conf
+ENV NATS_DOCKERIZED 1
+ENV NATS_SERVER 2.1.0
+ENV NATS_SERVER_DOWNLOAD https://github.com/nats-io/nats-server/releases/download/v${NATS_SERVER}/nats-server-v${NATS_SERVER}-windows-amd64.zip
 
-# Expose client, management, and cluster ports
-EXPOSE 4222 8222 6222
+RUN Set-PSDebug -Trace 2
+
+RUN Write-Host ('downloading from {0}' -f $env:NATS_SERVER_DOWNLOAD); \
+	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
+	Invoke-WebRequest -Uri $env:NATS_SERVER_DOWNLOAD -OutFile nats.zip; \
+	\
+	Write-Host 'extracting nats.zip'; \
+	Expand-Archive -Path 'nats.zip' -DestinationPath .; \
+	\
+	Write-Host 'copying binary'; \
+	Copy-Item nats-server-v*/nats-server.exe -Destination C:\\nats-server.exe; \
+	\
+	Write-Host 'cleaning up'; \
+	Remove-Item -Force nats.zip; \
+	Remove-Item -Recurse -Force nats-server-v*
 
-ENTRYPOINT ["nats-server"]
-CMD ["-c", "nats-server.conf"]
+COPY nats-server.conf C:\\nats-server.conf
+
+EXPOSE 4222 8222 6222
+ENTRYPOINT ["C:\\nats-server.exe"]
+CMD ["--config", "nats-server.conf"]
diff --git a/nats_windowsservercore/nats-server.exe b/nats_windowsservercore/nats-server.exe
deleted file mode 100755
index d5c9993..0000000
````

</details>

**Comment by tianon** [comment]
LGTM

Build test of #6766; 77b0fe77; `amd64` (`nats`):

```console

$ bashbrew build nats:2.1.0-linux
Building bashbrew/cache:60319626122024eebdda1c08cdc3f24f2787e1c8d04e76bffc3e885af463c04c (nats:2.1.0-linux)
Tagging nats:2.1.0-linux
Tagging nats:linux
Tagging nats:2.1.0
Tagging nats:latest

$ test/run.sh nats:2.1.0-linux
testing nats:2.1.0-linux
	'cve-2014--shellshock' [1/2]...passed
	'override-cmd' [2/2]...image does not appear to contain "echo" -- assuming single-binary image
passed


$ bashbrew build nats:2.1.0-windowsservercore-1809
skipping "nats:2.1.0-windowsservercore-1809" (due to architecture "amd64"; only "windows-amd64" supported)

$ test/run.sh nats:2.1.0-windowsservercore-1809
testing nats:2.1.0-windowsservercore-1809
	image has no tests...skipping


$ bashbrew build nats:2.1.0-nanoserver-1809
skipping "nats:2.1.0-nanoserver-1809" (due to architecture "amd64"; only "windows-amd64" supported)

$ test/run.sh nats:2.1.0-nanoserver-1809
testing nats:2.1.0-nanoserver-1809
	image has no tests...skipping


$ bashbrew build nats:2.1.0-windowsservercore-1803
skipping "nats:2.1.0-windowsservercore-1803" (due to architecture "amd64"; only "windows-amd64" supported)

$ test/run.sh nats:2.1.0-windowsservercore-1803
testing nats:2.1.0-windowsservercore-1803
	image has no tests...skipping


$ bashbrew build nats:2.1.0-nanoserver-1803
skipping "nats:2.1.0-nanoserver-1803" (due to architecture "amd64"; only "windows-amd64" supported)

$ test/run.sh nats:2.1.0-nanoserver-1803
testing nats:2.1.0-nanoserver-1803
	image has no tests...skipping


$ bashbrew build nats:2.1.0-windowsservercore-ltsc2016
skipping "nats:2.1.0-windowsservercore-ltsc2016" (due to architecture "amd64"; only "windows-amd64" supported)

$ test/run.sh nats:2.1.0-windowsservercore-ltsc2016
testing nats:2.1.0-windowsservercore-ltsc2016
	image has no tests...skipping

```

**Comment by tianon** [comment]
Regarding the binary artifacts, is it worth considering making the official release binaries fully static for later releases? (That would resolve the issues using it in a `FROM scratch` image, and have the benefit of making them even more portable for users/consumers who don't want to use a Docker image but perhaps have odd environments they want to run them in.)

Regarding `SharedTags`, you'll probably want to add the other `nanoserver` version as well so that Windows can do automatic version detection for users who aren't using `--isolation hyperv`.  Additionally, it might be worth considering turning the bare `nanoserver` and `windowsservercore` tags into `SharedTags` so that they can work similarly.

(See https://github.com/docker-library/official-images/blob/master/library/hello-world for a decently simple example of what I mean there if it's not clear. :+1:)

**Comment by variadico** [comment]
W00t. Thanks for the feedback. Created another PR: https://github.com/docker-library/official-images/pull/6812

I'll start the conversation about the fully static build on our side. It _would_ be nice to only have 1 way of building NATS.

**Pull request merged**